packages/go_router/lib/src/delegate.dart

```diff
@@ -42,22 +43,57 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
   /// Set to true to disable creating history entries on the web.
   final bool routerNeglect;
 
-  final GlobalKey<NavigatorState> _key = GlobalKey<NavigatorState>();
-
-  RouteMatchList _matches = RouteMatchList.empty();
+  RouteMatchList _matchList = RouteMatchList.empty();
   final Map<String, int> _pushCounts = <String, int>{};
+  final RouteConfiguration _configuration;
+
+  @override
+  Future<bool> popRoute() async {
+    // Iterate backwards through the RouteMatchList until seeing a GoRoute
+    // with a non-null parentNavigatorKey or a ShellRoute with a non-null parentNavigatorKey
+    // and pop from that Navigator instead of the root.
+    final int matchCount = _matchList.matches.length;
+    for (int i = matchCount - 1; i >= 0; i--) {
+      final RouteMatch match = _matchList.matches[i];
+      final RouteBase route = match.route;
+
+      // If this is a ShellRoute, then pop one of the subsequent GoRoutes, if
+      // there are any.
+      if (route is ShellRoute && (matchCount - i) > 2) {
```

Comment:
I think this should be assert((matchCount - i) > 2) ?

Reply:
If `matchCount - i` is greater than 2, that means that we can pop from that ShellRoute's navigator. It's valid for the ShellRoute's navigator to have a single page, in that case we wouldn't pop from this Navigator but one of the ancestor Navigator's instead. This seems to get the correct android back button behavior right.

Reply:
Isn't this already the ShellRoute's Navigator? We call maybePop on line 64

Reply:
Consider this case:
MatchList = [
  GoRoute1()
  ShellRoute()
  GoRouter()
]

The current code will skip the ShellRoute because (matchCount - i) <= 2, and call the maybePop on root navigator which will exit the app. If there is a pageless-route or open drawer in the last GoRouter it will get ignored.

The correct behavior is call maybePop at the first shell route to pop any pageless route or any local history entry that is not visible to flutter, and only call root navigator if the maybepop return false.

Reply:
I'm seeing a bigger issue here, regardless of what we do in this method we are removing the last match from the list of matches. Do you know why that might be happening? I'm going to see if I can fix that first.

Reply:
This should be fixed in d35a25b736d3b370953a11e10a7457fba8005b95